conmon-rs/server/src/pty.rs

```diff
@@ -0,0 +1,222 @@
+            self.0
+                .ok_or_else(|| io::Error::new(ErrorKind::NotFound, "stream not set"))?,
+            buf,
+        )
+        .map_err(io::Error::from)
+    }
+}
+
+impl Write for Stream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        unistd::write(
+            self.0
+                .ok_or_else(|| io::Error::new(ErrorKind::NotFound, "stream not set"))?,
+            buf,
+        )
+        .map_err(io::Error::from)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Pty {
+    /// Create a new Pty instance.
+    pub fn new(with_stdin: bool) -> Result<Self> {
+        debug!("Creating new pty");
+        let (main_stdin, worker_stdin) = if with_stdin {
+            let (x, y) = unistd::pipe2(OFlag::O_CLOEXEC).context("create stdin pipe")?;
+            (x.into(), y.into())
+        } else {
+            (Stream::default(), Stream::default())
+        };
+        let (main_stdout, worker_stdout) =
+            unistd::pipe2(OFlag::O_CLOEXEC).context("create stdout pipe")?;
+        let (main_stderr, worker_stderr) =
+            unistd::pipe2(OFlag::O_CLOEXEC).context("create stderr pipe")?;
+        Ok(Self {
+            main_stdin,
+            main_stderr: main_stderr.into(),
+            main_stdout: main_stdout.into(),
+            worker_stdin,
+            worker_stderr: worker_stderr.into(),
+            worker_stdout: worker_stdout.into(),
+        })
+    }
+
+    /// Create a Pty from a single raw file descriptor.
+    pub fn from_raw_fd(fd: RawFd) -> Result<Self> {
+        debug!("Creating pty from raw file descriptor");
+        let (main_stderr, worker_stderr) =
+            unistd::pipe2(OFlag::O_CLOEXEC).context("create stderr pipe")?;
+        Ok(Self {
+            main_stdin: fd.into(),
+            main_stderr: main_stderr.into(),
+            main_stdout: fd.into(),
+            worker_stdin: Stream::default(),
+            worker_stderr: worker_stderr.into(),
+            worker_stdout: Stream::default(),
+        })
+    }
+
+    /// Start the internal reading and writing threads.
+    pub fn start(&mut self) -> Result<()> {
+        self.prepare().context("prepare pty")?;
+
+        let stdin = self.main_stdin();
+        let stdout = self.main_stdout();
+        let stderr = self.main_stderr();
+        thread::spawn(move || Self::read_loop(stdin, stdout, stderr));
+        Ok(())
+    }
+
+    fn prepare(&mut self) -> Result<()> {
+        debug!("Preparing pty setup");
+        let mode = Mode::from_bits_truncate(0o777);
+        const DEV_NULL: &str = "/dev/null";
+
+        if !self.worker_stdin().is_set() {
+            let fd = OpenOptions::new().read(true).open(DEV_NULL)?.into_raw_fd();
+            unistd::dup2(fd, libc::STDIN_FILENO).context("dup over stdin")?;
+            self.set_worker_stdin(fd.into());
+        } else {
+            stat::fchmod(libc::STDIN_FILENO, mode).context("chmod stdin")?
+        }
```

Comment:
is a helper function here possible?

Reply:
Yeah, it would make things harder to read since there are slight differences between the arguments used in the functions. I decided to leave it as-is for now.